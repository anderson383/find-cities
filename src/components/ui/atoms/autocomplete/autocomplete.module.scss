.error {
  margin: 0;
  color: red;
  font-weight: 500;
}

.content {
  position: relative;
  width: 100%;
  .input {
    padding: 1rem;
    padding-left: 50px;
    min-width: 350px;
    border-radius: .5rem;
    border: none;
    box-sizing: border-box;
    font-size: 1rem;
    width: 100%;
    box-shadow: 0 3px 7px 0 rgba(0, 0, 3, 0.30588);
  }

  .inputSearch {
    box-sizing: border-box;
    padding-left: 3.5rem;
    background-color: #f1f1f1;
  }

  button {
    position: absolute;
    top: 0;
    bottom: 0;
    margin: auto 0;
    left: 1rem ;
    border: none;
    background: transparent;
    width: max-content;
    height: max-content;
  }

  .icon_search, .icon_delete {
    position: absolute;
    margin: auto 0;
    bottom: 0;
    top: 0;
  }
  .icon_search {
    left: 8px;
    width: 32px;
  }

  .icon_delete{
    cursor: pointer;
    position: absolute;
    right: 8px;
  }
}

.dropdown_options{
  margin: 8px 0;
  background-color: #ffff;
  z-index: 1;
  position: fixed;
  display: flex;
  box-sizing: border-box;
  overflow-y: auto;
  max-height: 350px;
  // overflow-x: hidden;
  flex-direction: column;
  box-shadow: 0 3px 7px 0 rgba(0, 0, 3, 0.30588);
  border-radius: .5rem;
  list-style: none;
  padding-left: 0;

  .dropdown_option {
    cursor: pointer;
    padding: 12px 32px;
  }

  .dropdown_option:hover {
    background-color: #f1f1f1;
  }

  .selected {
    background-color: #e1e1e1;
  }
  .currentHover {
    background-color: #efefef;
  }
}