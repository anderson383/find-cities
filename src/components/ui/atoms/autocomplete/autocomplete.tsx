
import {
  IconMapSearch, IconX
} from '@tabler/icons-react';
import {
  motion, Variants
} from 'framer-motion';
import React, {
  Dispatch, SetStateAction, useEffect, useRef, useState
} from 'react';
import scrollToSelected, { handleInteractiveKeyDown } from '../../../../helpers/inputs';
import _ from 'lodash';
import ReactDOM from 'react-dom';
import styles from './autocomplete.module.scss';
import { useWindowSize } from '../../../../hooks/use-window';

interface InputTextProps<T = any> {
  name:string
  placeholder?: string,
  value: T;
  setCurrentValue?: Dispatch<SetStateAction<T>>;
  onChangeText: Dispatch<SetStateAction<string>>;
  valueText: string
  onScroll?: (ref:any) => void
  items: T[]
  itemLabel?: string
}

export const variantsAutocomplete: Variants = {
  open: {
    clipPath: 'inset(0% 0% 0% 0% round 10px)',
    transition: {
      type: 'spring',
      bounce: 0,
      duration: 0.7,
      delayChildren: 0.3,
      staggerChildren: 0.05
    }
  },
  closed: {
    clipPath: 'inset(10% 50% 90% 50% round 10px)',
    transition: {
      type: 'spring',
      bounce: 0,
      duration: 0.3
    }
  }
};

export const itemVariants: Variants = {
  open: {
    opacity: 1,
    y: 0,
    transition: {
      type: 'spring',
      stiffness: 300,
      damping: 24
    }
  },
  closed: {
    opacity: 0,
    y: 20,
    transition: { duration: 0.2 }
  }
};

export const AutoComplete: React.FC<InputTextProps> = ({
  name,
  placeholder,
  setCurrentValue,
  value: currentValue,
  onChangeText,
  valueText,
  onScroll,
  items: itemsList,
  itemLabel
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const selectRef = useRef<HTMLDivElement>(null);
  const dropdownRef = React.useRef<any>(null);
  const [hoverItem, setHoverItem] = useState<any>(null);
  const optionsRef = useRef<any>(null);
  const [popoverPosition, setPopoverPosition] = useState({
    top: 0,
    left: 0,
    width: 0,
    height: 0
  });

  const portalContainer = document.getElementById('portal-root');

  const { width: widthScreen } = useWindowSize();

  const onReziseScreen = () => {
    if (dropdownRef.current) {
      const rect = dropdownRef.current.getBoundingClientRect();

      setPopoverPosition({
        width: rect.width,
        top: rect.bottom,
        left: rect.left,
        height:
          window.innerHeight - rect.bottom > 128 ? null : window.innerHeight - rect.bottom - 16
      });
    }
  };

  useEffect(() => {
    scrollToSelected({ optionsRef });
  }, [isOpen, hoverItem]);

  useEffect(() => {
    setHoverItem(currentValue);
    onReziseScreen();
  }, [widthScreen, isOpen]);

  const handleInteractiveKeyDownData = (e: React.KeyboardEvent<HTMLDivElement>) => {
    handleInteractiveKeyDown({
      e,
      isOpen,
      setIsOpen,
      onReziseScreen,
      hoverItem,
      setHoverItem,
      handleOptionClick,
      optionSelect: currentValue,
      optionElements: itemsList
    });
  };

  const handleOptionClick = (
    option: any,
    id?: string
  ) => {
    setIsOpen(false);
    setHoverItem(option);
    setCurrentValue(option);
  };

  return (
    <>
      <motion.div className={styles.content} ref={selectRef} animate={isOpen ? 'open' : 'closed'} onKeyDown={e => handleInteractiveKeyDownData(e)} >
        <div>
          <IconMapSearch role='icon-search-map' stroke={1.5} className={styles.icon_search} />
          <IconX role='icon-clear' stroke={1.5} className={styles.icon_delete} onClick={() => onChangeText('')} />
          <motion.input
            onClick={() => setIsOpen(!isOpen)}
            name={name}
            className={`${ styles.input }`}
            type={'text'}
            onChange={e => {
              setIsOpen(true);
              onChangeText(e.target.value);
            }}
            value={valueText}
            placeholder={placeholder}
            ref={dropdownRef}
          />
        </div>
        {
          ReactDOM.createPortal(
            <motion.ul
              className={styles.dropdown_options}
              style={{
                width: popoverPosition.width,
                top: popoverPosition.top,
                left: popoverPosition.left,
                height: popoverPosition.height,
                pointerEvents: isOpen ? 'auto' : 'none'
              }}
              ref={optionsRef}
              variants={variantsAutocomplete}
              onScroll={() => onScroll(optionsRef)}
            >
              {
                itemsList.map((item, index) => (
                  <motion.li
                    key={'list-autocomplete-' + index}
                    variants={itemVariants}
                    className={
                      `option ${ styles.dropdown_option } ${ _.isEqual(currentValue, item) ? styles.selected : '' } ${ _.isEqual(item, hoverItem) ? `selectedHover ${ styles.currentHover }` : '' }`
                    }
                    onClick={() => {
                      handleOptionClick(item);
                    }}
                  >
                    {item[itemLabel]}
                  </motion.li>
                ))
              }
            </motion.ul>,
            portalContainer
          )
        }
      </motion.div>
    </>
  );
};
