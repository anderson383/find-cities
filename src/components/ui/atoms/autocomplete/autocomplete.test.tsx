
import {
  fireEvent, render, screen
} from '@testing-library/react';
import { act } from 'react';
import { AutoComplete } from './autocomplete';
import { unmountComponentAtNode } from 'react-dom';

describe('Test component autocomplete', () => {
  let containerToast:any = null;
  const mockList = [
    {
      'country': 'US',
      'name': 'University Park',
      'lat': '32.28343',
      'lng': '-106.75334'
    },
    {
      'country': 'US',
      'name': 'Upper Fruitland',
      'lat': '36.71584',
      'lng': '-108.31424'
    },
    {
      'country': 'US',
      'name': 'Vado',
      'lat': '32.11176',
      'lng': '-106.6625'
    }
  ];

  const onChangeTextMock = jest.fn();
  const setCurrentValueMock = jest.fn();
  const onScrollMock = jest.fn();

  afterEach(() => {
    unmountComponentAtNode(containerToast);
    containerToast?.remove();
    containerToast = null;
  });

  beforeEach(() => {
    containerToast = document.createElement('div');
    containerToast.setAttribute('id', 'portal-root');

    document.body.appendChild(containerToast);
  });

  const setup = () => {
    return render(
      <AutoComplete
        items={mockList}
        name='Search'
        setCurrentValue={setCurrentValueMock}
        onChangeText={onChangeTextMock}
        value={''}
        valueText=''
        itemLabel='name'
        onScroll={onScrollMock}
        placeholder='Buscar prueba'
      />
    );
  };

  test('Should render component', async () => {
    // Arrange
    const {asFragment} = setup();

    // Act

    await act(() => Promise.resolve());

    // Assert
    expect(asFragment()).toMatchSnapshot();
  });

  test('Should write in input', async () => {
    // Arrange
    setup();
    const inputSearch = screen.getByPlaceholderText('Buscar prueba');

    // Act

    fireEvent.click(inputSearch);
    fireEvent.change(inputSearch, { target: { value: 'University' } });
    await act(() => Promise.resolve());

    // Assert
    expect(onChangeTextMock).toHaveBeenCalledWith('University');
  });

  test('Should clear text in input', async () => {
    // Arrange
    setup();
    const iconClear = screen.getByRole('icon-clear');
    const inputSearch = screen.getByPlaceholderText('Buscar prueba');

    // Act

    fireEvent.change(inputSearch, { target: { value: 'University' } });
    fireEvent.click(iconClear);
    await act(() => Promise.resolve());

    // Assert
    expect(onChangeTextMock).toHaveBeenCalledWith('');
  });

  test('Should press key for div', async () => {
    // Arrange
    const {container} = setup();
    const {firstChild} = container;

    fireEvent.keyDown(firstChild, { key: 'Enter'});

    // Act
    fireEvent.keyDown(firstChild, { key: 'ArrowDown' });
    fireEvent.keyDown(firstChild, { key: 'ArrowDown' });
    fireEvent.keyDown(firstChild, { key: 'ArrowDown' });

    fireEvent.keyDown(firstChild, { key: 'Enter' });
    await act(() => Promise.resolve());

    // Assert
    // Expect vado in array
    expect(setCurrentValueMock).toHaveBeenCalledWith(mockList[2]);
  });

  test('Should click in second option', async () => {
    // Arrange
    setup();

    const optionList = screen.getByText('Upper Fruitland');

    // Act
    fireEvent.click(optionList);
    await act(() => Promise.resolve());

    // Assert
    // Expect Upper Fruitland
    expect(setCurrentValueMock).toHaveBeenCalledWith(mockList[1]);
  });
});
