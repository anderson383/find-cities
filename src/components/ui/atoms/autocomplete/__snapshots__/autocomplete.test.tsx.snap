// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component autocomplete Should render component 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <div>
      <svg
        class="tabler-icon tabler-icon-map-search icon_search"
        fill="none"
        height="24"
        role="icon-search-map"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 18l-2 -1l-6 3v-13l6 -3l6 3l6 -3v7.5"
        />
        <path
          d="M9 4v13"
        />
        <path
          d="M15 7v5"
        />
        <path
          d="M18 18m-3 0a3 3 0 1 0 6 0a3 3 0 1 0 -6 0"
        />
        <path
          d="M20.2 20.2l1.8 1.8"
        />
      </svg>
      <svg
        class="tabler-icon tabler-icon-x icon_delete"
        fill="none"
        height="24"
        role="icon-clear"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6l-12 12"
        />
        <path
          d="M6 6l12 12"
        />
      </svg>
      <input
        class="input"
        name="Search"
        placeholder="Buscar prueba"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
